core/common/tarutil.cpp

```diff
@@ -38,6 +39,115 @@ namespace fc::common {
+      close(p[1]);
+      close(p[0]);
+    };
+    auto _ = gsl::finally([&]() { clear(); });
+
+    auto a = ffi::wrap(archive_write_new(), archive_write_free);
+    archive_write_set_format_v7tar(a.get());
+    int r = archive_write_open_fd(a.get(), p[1]);
+    if (r < ARCHIVE_OK) {
+      if (r < ARCHIVE_WARN) {
+        logger->error("Zip tar: {}", archive_error_string(a.get()));
+        return TarErrors::kCannotZipTarArchive;
+      } else {
+        logger->warn("Zip tar: {}", archive_error_string(a.get()));
+      }
+    }
+    std::function<outcome::result<void>(const fs::path &, const fs::path &)>
+        zipDir = [&](const fs::path &absolute_path,
+                     const fs::path &relative_path) -> outcome::result<void> {
+      struct archive_entry *entry;
+      struct stat st {};
+      char buff[8192];
+      fs::directory_iterator dir_iter(absolute_path), end;
+
+      while (dir_iter != end) {
+        if (fs::is_directory(dir_iter->path())) {
+          if (fs::directory_iterator(dir_iter->path())
+              != fs::directory_iterator()) {
+            OUTCOME_TRY(zipDir(dir_iter->path(),
+                               relative_path / dir_iter->path().filename()));
+          } else {
+            stat(dir_iter->path().c_str(), &st);
+            entry = archive_entry_new();  // Note 2
+            archive_entry_set_pathname(
+                entry, (relative_path / dir_iter->path().filename()).c_str());
+            archive_entry_set_size(entry, st.st_size);  // Note 3
+            archive_entry_set_filetype(entry, AE_IFDIR);
+            archive_entry_set_perm(entry, 0644);
+            r = archive_write_header(a.get(), entry);
+            if (r < ARCHIVE_OK) {
+              if (r < ARCHIVE_WARN) {
+                logger->error("Zip tar: {}", archive_error_string(a.get()));
+                return TarErrors::kCannotZipTarArchive;
+              } else {
+                logger->warn("Zip tar: {}", archive_error_string(a.get()));
+              }
+            }
+            archive_entry_free(entry);
+          }
+        } else if (fs::is_regular_file(dir_iter->path())) {
+          if (stat(dir_iter->path().c_str(), &st) < 0) {
+            return TarErrors::kCannotZipTarArchive;
+          }
+          entry = archive_entry_new();
+          archive_entry_set_pathname(
+              entry, (relative_path / dir_iter->path().filename()).c_str());
+          archive_entry_set_size(entry, st.st_size);
+          archive_entry_set_filetype(entry, AE_IFREG);
+          archive_entry_set_perm(entry, 0644);
+          r = archive_write_header(a.get(), entry);
+          if (r < ARCHIVE_OK) {
+            if (r < ARCHIVE_WARN) {
+              logger->error("Zip tar: {}", archive_error_string(a.get()));
+              return TarErrors::kCannotZipTarArchive;
+            } else {
+              logger->warn("Zip tar: {}", archive_error_string(a.get()));
+            }
+          }
```

Comment:
Duplicate code for regular file and empty directory

Reply:
Fixed